warn about potential issues of passing bad style prop to DOM elements

bobbyrenwick opened the pull request:
This adds a warning whenever there's a potentially dangerous spread of `this.props` followed by a `style` prop, on either a DOM component, or a variable. We need this because the `style` prop would previously override any `style` prop that came from the parent component, and Radium would resolve that into something the DOM can handle. Because glamor spreads data attributes, we no longer override the style prop, so in some cases we're causing errors like [TypeError: CSS2Properties doesn't have an indexed property setter for '0'](https://medium.com/@abhaytalreja/react-typeerror-css2properties-doesnt-have-an-indexed-property-setter-for-0-4acea43facd5).

We could make this more intelligent by seeing if there's a child of the style prop that mentions `this.props.style` in any way but I think that it's better to warn more readily, than to potentially mess up!

bobbyrenwick committed:
warn about potential issues of passing bad style prop to DOM elements

andrinealver commented on transforms/__testfixtures__/radium-to-glamor/this-props-spread.input.js:
```diff
@@ -0,0 +1,44 @@
+                    {...this.props}
+                    otherProp={true}
+                >
+                  NO DANGER BECAUSE NO STYLE
```
It's not necessarily the case that there's no danger when there's not style props getting passed to the DOM. It can still be part of the spread props.

bobbyrenwick replied:
How right you are - I'll make a fix now

bobbyrenwick committed:
Merge branch 'master' into this-props-spread-warning

Conflicts:
	transforms/__tests__/radium-to-glamor-test.js

bobbyrenwick committed:
change logic behind spread props warnings to be any DOM node with spread

The original logic was broken - it doesn't matter whether there's a style prop afterwards because the spread still can still contain the original `style` prop.

andrinealver commented on transforms/__testfixtures__/radium-to-glamor/this-props-spread.output.js:
```diff
@@ -1,6 +1,8 @@
 // TODO_RADIUM_TO_GLAMOR - JSX refers to ComponentClass, which is a variable. So wanted behaviour unknown.
-// TODO_RADIUM_TO_GLAMOR - In JSX the props for component, ComponentClass, contains a spread which references this.props, followed by a style prop. It's not safe to assume that this is safe because the style prop has been removed
-// TODO_RADIUM_TO_GLAMOR - In JSX the props for component, div, contains a spread which references this.props, followed by a style prop. It's not safe to assume that this is safe because the style prop has been removed
+// TODO_RADIUM_TO_GLAMOR - In JSX the props for component, ComponentClass, contains a spread which references this.props. It's not safe to assume that this is safe because it might contain a style prop that'll no longer be reconciled by Radium
```
This is a lot of comments.. What do you think about adding the component names / elements to a list and show that instead? So that it'd be `// TODO_RADIUM_TO_GLAMOR - In JSX the props for component, div, li, OtherClass and ComponentClass contains a spread which references this.props. It's not safe to assume that this is safe because it might contain a style prop that'll no longer be reconciled by Radium` ?

bobbyrenwick commented:
@andrinealver please could you take another look at this badboy?
